README.md
# eive
An R package for Errors-in-variables estimation in linear regression

## Installation

### Install stable version from CRAN

```R
install.packages("eive")
```


### Install development version 

Please install ```devtools``` package before installing ```eive```:

```R
install.packages("devtools")
```

then install the package from the github repo using

```R
devtools::install_github(repo = "https://github.com/jbytecode/eive") 
```

# The Problem 

Suppose the linear regression model is 

$$
y = \beta_0 + \beta_1 x^* + \varepsilon
$$

where $y$ is n-vector of the response variable, $\beta_0$ and $\beta_1$ are unknown regression parameteres, $\varepsilon$ is the iid. error term, $x^*$ is the unknown n-vector of the independent variable, and $n$ is the number of observations.

We call $x^*$ unknown because in some situations the true values of the variable cannot be visible or directly observable, or observable with some measurement error. Now suppose that $x$ is the observable version of the true values and it is defined as 

$$
x = x^* + \delta
$$

where $\delta$ is the measurement error and $x$ is the erroneous version of the true $x^*$. If the estimated model is 

$$
\hat{y} = \hat{\beta_0} + \hat{\beta_1}x 
$$

then the ordinary least squares (OLS) estimates are no longer unbiased and even consistent. 

Eive-cga is an estimator devised for this problem. The aim is to reduce the errors-in-variable bias with some cost of increasing the variance. At the end, the estimator obtains lower Mean Square Error (MSE) values defined as

$$
MSE(\hat{\beta_1}) = Var(\hat{\beta_1}) + Bias^2(\hat{\beta_1})
$$

for the Eive-cga estimator. For more detailed comparisons, see the original paper given in the Citation part. 

# Usage 

For the single variable case 

```R 
> eive(dirtyx = dirtyx, y = y, otherx = nothing) 
```

and for the multiple regression 

```R 
> eive(dirtyx = dirtyx, y = y, otherx = matrixofotherx) 
```

and for the multiple regression with formula object 

```R 
> eive(formula = y ~ x1 + x2 + x3, dirtyx.varname = "x", data = mydata) 
```

Note that the method assumes there is only one erroneous variable in the set of independent variables.

### Citation 
```bibtex
@article{satman2015reducing,
  title={Reducing errors-in-variables bias in linear regression using compact genetic algorithms},
  author={Satman, M Hakan and Diyarbakirlioglu, Erkin},
  journal={Journal of Statistical Computation and Simulation},
  volume={85},
  number={16},
  pages={3216--3235},
  year={2015},
  doi={10.1080/00949655.2014.961157}
  publisher={Taylor \& Francis}
}
```

CHANGELOG.md
# v3.1.3 (Upcoming version)

# v3.1.2
- Add eive.cga.formula for lm() compatible regression settings.
- Fix documentation

# v3.1.1
- Enhance document for multivariate eive (eivem)
  

## Version 3.1.0 

- Errors-in-variables with multiple response variables.
- eive.cga() now returns 2 new items: $cleanx and $measurementerror which are also accessable using the keys $proxy$fitted.values and $proxy$residuals, respectively.
- eivem() added for multiple y values
- new tests added and can be triggered using devtools:::test()
- updated docs




tests/testthat.R
library(testthat)
library(eive)

test_check("eive")

tests/testthat/test-cga.R
library("testthat")


test_that("CGA - All 10 bits are 1s", {
    f <- function(bits) {
        return(-sum(bits))
    }
    n <- 10
    result <- cga(n, 100, f)
    expect_equal(result, rep(1, n))
})

test_that("CGA - All 100 bits are 1s", {
    f <- function(bits) {
        return(-sum(bits))
    }
    n <- 100
    result <- cga(n, 100, f)
    expect_equal(result, rep(1, n))
})

test_that("CGA - All 100 bits are 0s", {
    f <- function(bits) {
        return(sum(bits))
    }
    n <- 100
    result <- cga(n, 100, f)
    expect_equal(result, rep(0, n))
})

test_that("CGA - Generate Chromosome", {
    probs <- rep(0.5, 1000)
    v <- rep(0, 1000)
    result <- cga_generate_chromosome(probs, v)
    
    # Function returns NULL
    expect_null(result)
    
    # Not all of the values are same
    expect_false(all(v == 0))
    expect_false(all(v == 1))
})

tests/testthat/test-eive.R
library(testthat)
library(eive)

test_that("eive.cga classical example", {

    tol <- 0.01

    euclidean <- function(u, v) {
        return(sqrt(sum((u - v)^2)))
    }

    require("eive")

    set.seed(12345)

    n <- 30

    clean_x <- rnorm(n, mean = 10, sd = sqrt(7))

    delta_x <- rnorm(n, mean = 0, sd = sqrt(3))

    e <- rnorm(n, mean = 0, sd = sqrt(5))

    y <- 20 + 10 * clean_x + e

    dirty_x <- clean_x + delta_x

    result <- eive.cga(dirtyx = dirty_x, y = y, numdummies = 10)

    par_real <- c(20, 10)
    par_ols <- result$ols$coefficients
    par_eive <- result$eive$coefficients

    dist1 <- euclidean(par_real, par_ols)
    dist2 <- euclidean(par_real, par_eive)

    expect_true(dist2 < dist1)

    expect_equal(as.vector(par_eive), c(23.863, 9.229), tolerance = tol)
    expect_equal(as.vector(par_ols), c(63.590, 5.533), tolerance = tol)

    L <- length(result$measurementerror)
    for (i in 1:L){
        expect_equal(result$proxy$residuals[i], result$measurementerror[i])
    }

    result_names <- names(result)
    expect_true("ols" %in% result_names)
    expect_true("eive" %in% result_names)
    expect_true("proxy" %in% result_names)
    expect_true("cleanedx" %in% result_names)
    expect_true("measurementerror" %in% result_names)
})



R/eivem.R
library(compiler)


#' Performs CGA based errors-in-variables correction for a given set of variables
#' in case of multiple Y variables are provided.
#' @description  A single independent variable is supposed to be measured subject to error.
#'               This functions is the multivariate version of the classical algorithm.
#'               Additional response variables are used to get better estimates.
#'
#' @param dirtyx Vector of independent variable that is measured with error.
#' @param otherx Matrix of other independent variables. If the model has
#'        a single independent variable, it is NULL by default.
#' @param y Matrix of response variables. Y_i is placed in the ith row of the matrix.
#' @param numdummies Number of dummy variables used in auxiliary regression. Default is 10.
#' @param popsize Population size parameter for compact genetic algorithm. Default is 20.
#' 1/popsize is the mutation rate.
#' @return A list() of regression equations.
#' @slot ols List of lm objects calculated using original values
#' @slot eive List of lm objects calculated using the predicted variable by eive
#' @slot proxy lm object of proxy regression obtained by genetic search.
#' @slot cleanedx Error-free estimate of the x variable (dirtyx)
#'       that is measured with error.
#' @slot measurementerror Estimate of the measurement error.
#' @export 
#' @examples
#' # Creating an artificial data
#' 
#' # Loading required package
#' require("eive")
#' 
#' # Setting random number generator seed to 12345
#' # so each time the script runs, same numbers will
#' # be generated
#' set.seed(12345)
#' 
#' # Number of observations is set to 30
#' n <- 30
#' 
#' # Unobserved X values are drawn from a Normal distribution
#' # with mean 10 and variance 7
#' clean_x1 <- rnorm(n, mean = 10, sd = sqrt(7))
#' clean_x2 <- rnorm(n, mean = 10, sd = sqrt(7))
#' 
#' # Measurement error values are dranw from a Normal distribution
#' # with mean 0 and variance 3
#' delta_x1 <- rnorm(n, mean = 0, sd = sqrt(3))
#' 
#' # Error term of regression. Normally distributed with mean 0 and
#' # variance 5
#' e1 <- rnorm(n, mean = 0, sd = sqrt(5))
#' e2 <- rnorm(n, mean = 0, sd = sqrt(5))
#' 
#' # Generating Y values using the linear model
#' # In this model, intercept is 20 and slope is 10.
#' y1 <- 20 + 10 * clean_x1 + 10 * clean_x2 + e1
#' y2 <- 10 + 5 * clean_x1 + 5 * clean_x2 + e2
#' 
#' # Generating observed X values by adding measurement errors
#' # to unobserved X
#' dirty_x1 <- clean_x1 + delta_x1
#' 
#' # Performs a genetic search to find dummy variables that
#' # used in two stage least squares.
#' # Please un-comment the line below
#' result <- eivem(dirtyx = dirty_x1, otherx = clean_x2, y = cbind(y1, y2), numdummies = 10)
#' 
#' # Print the result
#' # Please un-comment the line below
#' # print(result)
#' 
#' ########################################### OUTPUT #############################################
#' #> result
#' # $ols
#' # $ols[[1]]
#' #
#' # Call:
#' # lm(formula = y[, reg.index] ~ dirtyx + otherx)
#' #
#' # Coefficients:
#' # (Intercept)       dirtyx       otherx
#' #      54.141        6.067       10.137
#' #
#' #
#' # $ols[[2]]
#' #
#' # Call:
#' # lm(formula = y[, reg.index] ~ dirtyx + otherx)
#' #
#' # Coefficients:
#' # (Intercept)       dirtyx       otherx
#' #      24.814        3.205        5.089
#' #
#' #
#' #
#' # $eive
#' # $eive[[1]]
#' #
#' # Call:
#' # lm(formula = y[, reg.index] ~ ols_proxy$fitted.values + otherx)
#' #
#' # Coefficients:
#' #             (Intercept)  ols_proxy$fitted.values                   otherx
#' #                  24.737                    9.727                    9.147
#' #
#' #
#' # $eive[[2]]
#' #
#' # Call:
#' # lm(formula = y[, reg.index] ~ ols_proxy$fitted.values + otherx)
#' #
#' # Coefficients:
#' #             (Intercept)  ols_proxy$fitted.values                   otherx
#' #                   8.313                    5.240                    4.552
#' #
#' #
#' 
#' # $proxy
#' #
#' # Call:
#' # lm(formula = dirtyx ~ matrix(best, nrow = n))
#' #
#' # Coefficients:
#' #              (Intercept)   matrix(best, nrow = n)1   matrix(best, nrow = n)2
#' #                 6.314397                 -0.211580                  1.729143
#' #  matrix(best, nrow = n)3   matrix(best, nrow = n)4   matrix(best, nrow = n)5
#' #                 1.994915                  0.947531                 -0.363107
#' #  matrix(best, nrow = n)6   matrix(best, nrow = n)7   matrix(best, nrow = n)8
#' #                 0.001768                  1.742553                 -0.023750
#' #  matrix(best, nrow = n)9  matrix(best, nrow = n)10
#' #                 0.134750                  2.324853
#' #
#' #
#' # $cleanedx
#' #         1         2         3         4         5         6         7         8
#' # 12.730307 12.130102 11.065586  9.795474 12.697138  6.450915 12.673388 10.516553
#' #         9        10        11        12        13        14        15        16
#' # 11.095771  7.981887 11.694464 14.841812 11.098755 12.290371  8.988344 12.704789
#' #        17        18        19        20        21        22        23        24
#' #  7.671861  9.477178 13.458999 10.964004 11.465852 14.591473  9.771724  6.239335
#' #        25        26        27        28        29        30
#' #  6.425397 15.031410  8.992839 12.808138 13.435249  9.799758
#' #
#' # $measurementerror
#' #          1          2          3          4          5          6          7
#' # -0.9220426 -2.5783644 -0.3964263  1.7585818 -2.1106159 -4.4345451  0.5319987
#' #          8          9         10         11         12         13         14
#' #  1.5127360 -0.9523682 -2.6583539 -1.9074299 -1.3927085 -1.9356982  3.1225578
#' #         15         16         17         18         19         20         21
#' #  1.4554922  1.0891572  1.4141792 -1.7600789  0.3310142  1.5952156  1.7146703
#' #         22         23         24         25         26         27         28
#' #  1.0669497 -2.0036393  3.9419318  1.0296643  2.9783401  0.8968531 -1.7001587
#' #         29         30
#' # -0.8864360  1.1995241
#' #
#' ######################################### END OF OUTPUT ##########################################
eivem <- function(dirtyx, otherx = NULL, y, numdummies = 10, popsize = 20) {
  ols_dirty <- list()
  ols_proxy <- NULL
  ols_best <- list()
  my_dim <- dim(y)
  n <- my_dim[1]
  yp <- my_dim[2]

  f <- function(d) {
    ols <- NULL
    m <- matrix(d, nrow = n)
    ols_proxy <- lm.fit(cbind(1, m), dirtyx)
    x_proxy <- ols_proxy$fitted.values
    square_sums <- 0
    for (reg.index in 1:yp) {
      if (is.null(otherx)) {
        ols <- lm.fit(cbind(1, x_proxy), y[, reg.index])
      } else {
        ols <- lm.fit(cbind(1, x_proxy, otherx), y[, reg.index])
      }
      square_sums <- square_sums + sum(ols$residuals^2)
    }
    return(square_sums)
  }

  g <- cmpfun(f)

  ga <- cga(evalFunc = g, chsize = n * numdummies, popsize = popsize)

  best <- as.integer(ga)

  for (reg.index in 1:yp) {
    if (is.null(otherx)) {
      ols_dirty[[reg.index]] <- lm(y[, reg.index] ~ dirtyx)
    } else {
      ols_dirty[[reg.index]] <- lm(y[, reg.index] ~ dirtyx + otherx)
    }
    ols_proxy <- lm(dirtyx ~ matrix(best, nrow = n))
    if (is.null(otherx)) {
      ols_best[[reg.index]] <- lm(y[, reg.index] ~ ols_proxy$fitted.values)
    } else {
      ols_best[[reg.index]] <- lm(
        y[, reg.index] ~ ols_proxy$fitted.values + otherx
      )
    }
  } # end of for loop

  cleanedx <- ols_proxy$fitted.values
  measurementerror <- ols_proxy$residuals

  result <- list(
    ols = ols_dirty,
    eive = ols_best,
    proxy = ols_proxy,
    cleanedx = cleanedx,
    measurementerror = measurementerror
  )

  return(result)
}
# End of function


src/ccga.cpp
#include <Rcpp.h>
using namespace Rcpp;

//' @name cga_generate_chromosome
//' @title Generate Chromosome
//' @description Generate a binary vector using a probability vector
//' 	This function is not directly called by user. CGAs (Compact genetic algorithms)
//' 	sample chromosomes using this probability vector. A probability vector
//' 	contains[P1, P2, ..., PN] and the function generates and returns a chromosome[B1, B2, ..., BN].
//' 	The probability of BK having the value of 1 is PK. So, it has more chance to have
//' 	[1, 1, 1, 0, 0] rather than [0, 0, 0, 1, 1] when the probability vector is
//'  	[0.9, 0.9, 0.9, 0.1, 0.1]. 
//' @param prob_vec Vector of probabilities
//' @param vect Vector of bits.
//' @return Mutates the vect. Returns null.
//' @export
// [[Rcpp::export]]
void cga_generate_chromosome(NumericVector prob_vec, NumericVector vect)
{
	int len_prob_vec = prob_vec.length();
	NumericVector unifs = runif(len_prob_vec);
	int i;
	for (i = 0; i < len_prob_vec; i++)
	{
		if (unifs[i] < prob_vec[i])
		{
			vect[i] = 1;
		}else{
			vect[i] = 0;
		}
	}
}

//' @name cga
//' @title Compact Genetic Algorithm
//' @description Performs a Compact Genetic Algorithm (CGA) search
//' 	for a given chromosome size, population size (mutation rate), 
//'     and an objective function. 
//' @param chsize Number of bits.
//' @param popsize Size of population. The value is used for mutating 
//'		the probability vector by 1/popsize. 
//' @param evalFunc Objective function.
//' @return Binary vector of size chsize.
//' @export
// [[Rcpp::export]]
NumericVector cga(int chsize, int popsize, Function evalFunc)
{
	NumericVector prob_vec = rep(0.5, chsize);
	NumericVector chromosome1(chsize);
	NumericVector chromosome2(chsize);
	NumericVector winner, loser;
	NumericVector cost1, cost2;
	int i, t;
	double mutation = 1.0 / (double)popsize;
	while (1)
	{
		cga_generate_chromosome(prob_vec, chromosome1);
		cga_generate_chromosome(prob_vec, chromosome2);
		cost1 = evalFunc(chromosome1);
		cost2 = evalFunc(chromosome2);
		winner = chromosome1;
		loser = chromosome2;
		if (cost2[0] < cost1[0])
		{
			winner = chromosome2;
			loser = chromosome1;
		}

		for (i = 0; i < chsize; i++)
		{
			if (winner[i] != loser[i])
			{
				if (winner[i] == 1)
				{
					prob_vec[i] = prob_vec[i] + mutation;
				}
				else
				{
					prob_vec[i] = prob_vec[i] - mutation;
				}
			}
		}

		t = 0;
		for (i = 0; i < chsize; i++)
		{
			if (prob_vec[i] <= 0.001 || prob_vec[i] >= 0.999)
			{
				t++;
			}
		}
		if (t == chsize)
		{
			break;
		}
	}
	return (prob_vec);
}


R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @name cga_generate_chromosome
#' @title Generate Chromosome
#' @description Generate a binary vector using a probability vector
#' 	This function is not directly called by user. CGAs (Compact genetic algorithms)
#' 	sample chromosomes using this probability vector. A probability vector
#' 	contains[P1, P2, ..., PN] and the function generates and returns a chromosome[B1, B2, ..., BN].
#' 	The probability of BK having the value of 1 is PK. So, it has more chance to have
#' 	[1, 1, 1, 0, 0] rather than [0, 0, 0, 1, 1] when the probability vector is
#'  	[0.9, 0.9, 0.9, 0.1, 0.1]. 
#' @param prob_vec Vector of probabilities
#' @param vect Vector of bits.
#' @return Mutates the vect. Returns null.
#' @export
cga_generate_chromosome <- function(prob_vec, vect) {
    invisible(.Call('_eive_cga_generate_chromosome', PACKAGE = 'eive', prob_vec, vect))
}

#' @name cga
#' @title Compact Genetic Algorithm
#' @description Performs a Compact Genetic Algorithm (CGA) search
#' 	for a given chromosome size, population size (mutation rate), 
#'     and an objective function. 
#' @param chsize Number of bits.
#' @param popsize Size of population. The value is used for mutating 
#'		the probability vector by 1/popsize. 
#' @param evalFunc Objective function.
#' @return Binary vector of size chsize.
#' @export
cga <- function(chsize, popsize, evalFunc) {
    .Call('_eive_cga', PACKAGE = 'eive', chsize, popsize, evalFunc)
}



src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <Rcpp.h>

using namespace Rcpp;

#ifdef RCPP_USE_GLOBAL_ROSTREAM
Rcpp::Rostream<true>&  Rcpp::Rcout = Rcpp::Rcpp_cout_get();
Rcpp::Rostream<false>& Rcpp::Rcerr = Rcpp::Rcpp_cerr_get();
#endif

// cga_generate_chromosome
void cga_generate_chromosome(NumericVector prob_vec, NumericVector vect);
RcppExport SEXP _eive_cga_generate_chromosome(SEXP prob_vecSEXP, SEXP vectSEXP) {
BEGIN_RCPP
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< NumericVector >::type prob_vec(prob_vecSEXP);
    Rcpp::traits::input_parameter< NumericVector >::type vect(vectSEXP);
    cga_generate_chromosome(prob_vec, vect);
    return R_NilValue;
END_RCPP
}
// cga
NumericVector cga(int chsize, int popsize, Function evalFunc);
RcppExport SEXP _eive_cga(SEXP chsizeSEXP, SEXP popsizeSEXP, SEXP evalFuncSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< int >::type chsize(chsizeSEXP);
    Rcpp::traits::input_parameter< int >::type popsize(popsizeSEXP);
    Rcpp::traits::input_parameter< Function >::type evalFunc(evalFuncSEXP);
    rcpp_result_gen = Rcpp::wrap(cga(chsize, popsize, evalFunc));
    return rcpp_result_gen;
END_RCPP
}


R/eive.R
library(compiler)

#' Generates simulated errors-in-variables regression data
#'
#' @param n Number of observations
#' @param e.sd Standard deviation of error term of regression
#' @param delta.sd Standard deviation of error in exploratory variable
#' @param seed Seed for random number generator. 12345 by default
#' @param useotherx Logical. If TRUE, an additional independent variable is added.
#' @return A matrix of variables.
#' @slot xdelta Errorenous X variable
#' @slot otherx Other X variable
#' @slot y Response variable
#' @export
generate.eive.data <- function(n,
                               e.sd,
                               delta.sd,
                               seed = 12345,
                               useotherx = FALSE) {
    set.seed(seed)
    e <- rnorm(n, 0, e.sd)
    x <- rnorm(n)
    x1 <- NULL
    y <- NULL
    data <- NULL
    if (useotherx) {
        x1 <- rnorm(n)
    }
    delta <- rnorm(n, 0, delta.sd)
    xdelta <- x + delta
    if (useotherx) {
        y <- 5 + 5 * x + 5 * x1 + e
    } else {
        y <- 5 + 5 * x + e
    }
    if (useotherx) {
        data <- data.frame(
            xdelta = xdelta, 
            xother = x1, 
            y = y
        )
    }else{
        data <- data.frame(
            xdelta = xdelta, 
            y = y
        )
    }
    return(data)
}


#' Performs CGA based errors-in-variables correction for given formula and data.
#' A single independent variable is supposed to be measured subject to error.
#'
#' @param formula Formula object.
#' @param data data.frame that holds the regression data.
#' @param dirtyx.varname String key value of the erroneous independent variable.
#' @param numdummies Number of dummy variables used in auxiliary regression.
#' @param popsize Population size parameter for compact genetic algorithm. 
#' 1/popsize is the mutation rate.
#' @return A list() of regression equations.
#' @slot ols lm object calculated using original values
#' @slot eive lm object calculated using the predicted variable by eive
#' @slot proxy lm object of proxy regression obtained by genetic search.
#' @slot cleanedx Error-free estimate of the x variable (dirtyx) 
#'       that is measured with error.
#' @slot measurementerror Estimate of the measurement error.
#' @export
#' @examples
#' set.seed(12345)
#' n <- 30
#' clean_x <- rnorm(n, mean = 10, sd = sqrt(7))
#' delta_x <- rnorm(n, mean = 0, sd = sqrt(3))
#' 
#' e <- rnorm(n, mean = 0, sd = sqrt(5))
#' y <- 20 + 10 * clean_x + e
#' 
#' dirty_x <- clean_x + delta_x
#' 
#' mydata <- data.frame(y = y, dirtyx = dirty_x)
#' 
#' result <- eive.cga.formula(
#'      formula = y ~ dirtyx,
#'      dirtyx.varname = "dirtyx",
#'      data = mydata,
#'      numdummies = 10
#' )
#' @seealso eive.cga
eive.cga.formula <- function(
    formula,
    data, 
    dirtyx.varname,
    numdummies = 10,
    popsize = 20
){
    if (!is.data.frame(data)) {
        stop("data should be in type of data.frame")
    }
    if (!is.language(formula)) {
        stop("formula should be in form of var ~ var + var ...")
    }
    if (!is.character(dirtyx.varname)) {
        stop("dirtyx.varname should be in type of string")
    }
    designmat <- as.data.frame(model.matrix(object = formula, data = data))
    responsevar <- model.frame(formula = formula, data = data)
    y <- responsevar[, 1]
    n <- length(y)
    p <- ncol(designmat)
    dirtyx <- as.vector(data[, dirtyx.varname])
    otherx <- designmat[, setdiff(names(designmat), dirtyx.varname)]

    if (is.vector(otherx)){
        otherx <- as.matrix(otherx)
    }
    has.intercept <- all.equal(otherx[, 1], rep(1, n))
    if (has.intercept) {
        if (ncol(otherx) > 1) {
            otherx <- as.matrix(otherx[, 2:(ncol(otherx))])
        }else{
            otherx <- NULL
        }
    }

    return(
        eive.cga(
            dirtyx = dirtyx, 
            otherx = otherx,
            y = y, 
            numdummies = numdummies, 
            popsize = popsize
        ))
}





#' Performs CGA based errors-in-variables correction for a given set of variables.
#' A single independent variable is supposed to be measured subject to error.
#'
#' @param dirtyx Vector of independent variable that is measured with error.
#' @param otherx Matrix of other independent variables. If the model has 
#'        a single independent variable, it is NULL by default.
#' @param y Vector of response variable
#' @param numdummies Number of dummy variables used in auxiliary regression.
#' @param popsize Population size parameter for compact genetic algorithm.
#' 1/popsize is the mutation rate.
#' @return A list() of regression equations.
#' @slot ols lm object calculated using original values
#' @slot eive lm object calculated using the predicted variable by eive
#' @slot proxy lm object of proxy regression obtained by genetic search.
#' @slot cleanedx Error-free estimate of the x variable (dirtyx)
#'       that is measured with error.
#' @slot measurementerror Estimate of the measurement error.
#' @export 
#' @examples 
#' # Creating an artificial data
#' # Loading required package
#' require("eive")
#' 
#' # Setting random number generator seed to 12345
#' # so each time the script runs, same numbers will
#' # be generated
#' set.seed(12345)
#' 
#' # Number of observations is set to 30
#' n <- 30
#' 
#' # Unobserved X values are drawn from a Normal distribution
#' # with mean 10 and variance 7
#' clean.x <- rnorm(n, mean = 10, sd = sqrt(7))
#' 
#' # Measurement error values are dranw from a Normal distribution
#' # with mean 0 and variance 3
#' delta.x <- rnorm(n, mean = 0, sd = sqrt(3))
#' 
#' # Error term of regression. Normally distributed with mean 0 and
#' # variance 5
#' e <- rnorm(n, mean = 0, sd = sqrt(5))
#' 
#' # Generating Y values using the linear model
#' # In this model, intercept is 20 and slope is 10.
#' y <- 20 + 10 * clean.x + e
#' 
#' # Generating observed X values by adding measurement errors
#' # to unobserved X
#' dirty.x <- clean.x + delta.x
#' 
#' # Performs a genetic search to find dummy variables that
#' # used in two stage least squares.
#' # Please un-comment the line below
#' # result <- eive.cga (dirtyx=dirty.x, y=y, numdummies=10)
#' 
#' # Print the result
#' # Please un-comment the line below
#' # print(result)
#' 
#' ########################################### OUTPUT #############################################
#' # $ols
#' #
#' # Call:
#' # lm(formula = y ~ dirtyx)
#' #
#' # Coefficients:
#' # (Intercept)       dirtyx
#' #     63.590        5.533
#' #
#' #
#' # $eive
#' #
#' # Call:
#' # lm(formula = y ~ ols.proxy$fitted.values)
#' #
#' # Coefficients:
#' #            (Intercept)  ols.proxy$fitted.values
#' #                 23.863                    9.229
#' #
#' #
#' # $proxy
#' #
#' # Call:
#' # lm(formula = dirtyx ~ matrix(best, nrow = n))
#' #
#' # Coefficients:
#' #              (Intercept)   matrix(best, nrow = n)1   matrix(best, nrow = n)2
#' #                 12.9321                   -0.6252                   -1.9923
#' # matrix(best, nrow = n)3   matrix(best, nrow = n)4   matrix(best, nrow = n)5
#' #                  0.7537                   -0.7076                   -0.5247
#' # matrix(best, nrow = n)6   matrix(best, nrow = n)7   matrix(best, nrow = n)8
#' #                 -0.9196                   -2.0802                   -0.9246
#' # matrix(best, nrow = n)9  matrix(best, nrow = n)10
#' #                 -0.6164                    1.9694
#' ######################################### END OF OUTPUT ##########################################
eive.cga <- function(dirtyx,
                     otherx = NULL,
                     y,
                     numdummies = 10,
                     popsize = 20) {
    ols_dirty <- NULL
    ols_proxy <- NULL
    ols_best <- NULL
    n <- length(y)
    f <- function(d) {
        ols <- NULL
        m <- matrix(d, nrow = n)
        ols_proxy <- lm.fit(cbind(1, m), dirtyx)
        x_proxy <- ols_proxy$fitted.values
        if (is.null(otherx)) {
            ols <- lm.fit(cbind(1, x_proxy), y)
        } else {
            ols <- lm.fit(cbind(1, x_proxy, otherx), y)
        }
        return(sum(ols$residuals^2))
    }

    ga <- cga(
        evalFunc = f,
        chsize = n * numdummies,
        popsize = popsize
    )
    
    best <- as.integer(ga)

    if (is.null(otherx)) {
        ols_dirty <- lm(y ~ dirtyx)
    } else {
        ols_dirty <- lm(y ~ dirtyx + otherx)
    }

    ols_proxy <- lm(dirtyx ~ matrix(best, nrow = n))

    if (is.null(otherx)) {
        ols_best <- lm(y ~ ols_proxy$fitted.values)
    } else {
        ols_best <- lm(y ~ ols_proxy$fitted.values + otherx)
    }

    cleanedx <- ols_proxy$fitted.values
    measurementerror <- ols_proxy$residuals
    
    result <- list(
      ols = ols_dirty,
      eive = ols_best,
      proxy = ols_proxy,
      cleanedx = cleanedx,
      measurementerror = measurementerror
    )
    
    return(result)
}


